src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1207 @@
+                                       *device_subhistograms_, *sync_counters_, device_histogram_outputs_);
+    histogram_fulldata_kernels_[i].set_args(*device_features_, device_feature_masks_, num_data_,
+                                        *device_data_indices_, num_data_, device_gradients_, device_hessians_,
+                                        *device_subhistograms_, *sync_counters_, device_histogram_outputs_);
+  }
+  Log::Info("%d dense features (%.2f MB) transfered to GPU in %f secs. %d sparse features.", 
+            dense_feature_map_.size(), ((dense_feature_map_.size() + (dword_features_ - 1)) / dword_features_) * num_data_ * sizeof(Feature4) / (1024.0 * 1024.0), 
+            end_time * 1e-3, sparse_feature_map_.size());
+  #if GPU_DEBUG >= 1
+  printf("Dense feature list (size %lu): ", dense_feature_map_.size());
+  for (i = 0; i < num_dense_features_; ++i) {
+    printf("%d ", dense_feature_map_[i]);
+  }
+  printf("\n");
+  printf("Sparse feature list (size %lu): ", sparse_feature_map_.size());
+  for (i = 0; i < num_features_ - num_dense_features_; ++i) {
+    printf("%d ", sparse_feature_map_[i]);
+  }
+  printf("\n");
+  #endif
+}
+
+void GPUTreeLearner::InitGPU(int platform_id, int device_id) {
+  // currently we don't use constant memory
+  int use_constants = 0;
+
+  // Get the max bin size, used for selecting best GPU kernel
+  max_num_bin_ = 0;
+  #if GPU_DEBUG >= 1
+  printf("bin size: ");
+  #endif
+  for (int i = 0; i < train_data_->num_features(); ++i) {
+    #if GPU_DEBUG >= 1
+    printf("%d, ", train_data_->FeatureNumTotalBin(i));
+    #endif
+    max_num_bin_ = std::max(max_num_bin_, train_data_->FeatureNumTotalBin(i));
+  }
+  #if GPU_DEBUG >= 1
+  printf("\n");
+  #endif
+  // initialize GPU
+  dev_ = boost::compute::system::default_device();
+  if (platform_id >= 0 && device_id >= 0) {
+    const std::vector<boost::compute::platform> platforms = boost::compute::system::platforms();
+    if ((int)platforms.size() > platform_id) {
+      const std::vector<boost::compute::device> platform_devices = platforms[platform_id].devices();
+      if ((int)platform_devices.size() > device_id) {
+        Log::Info("Using requested OpenCL platform %d device %d", platform_id, device_id);
+        dev_ = platform_devices[device_id];
+      }   
+    }   
+  }   
+  std::string kernel_source;
+  std::string kernel_name;
+  // determine which kernel to use based on the max number of bins
+  if (max_num_bin_ <= 16) {
+    kernel_source = kernel16_src_;
+    kernel_name = "histogram16";
+    device_bin_size_ = 16;
+    dword_features_ = 8;
+  }
+  else if (max_num_bin_ <= 64) {
+    kernel_source = kernel64_src_;
+    kernel_name = "histogram64";
+    device_bin_size_ = 64;
+    dword_features_ = 4;
+  }
+  else if ( max_num_bin_ <= 256) {
+    kernel_source = kernel256_src_;
+    kernel_name = "histogram256";
+    device_bin_size_ = 256;
+    dword_features_ = 4;
+  }
```

Comment:
Thanks for pointing out this! I think it should work. The code (line 564) checks FeatureNumTotalBin() for each feature, and the variable max_num_bin_ is the maximum number bins. FeatureNumTotalBin() returns the attribute num_total_bin_ of each FeatureGroup. So max_num_bin_ is not the command line parameter max_bin. It depends on the real number of bins used. 

If a feature only contains 0/1 values, then I think max_num_bin_ will be 2. If all features are like this, then max_num_bin_ will be set to 2, and dword_features_ will be set to 8. In this case, all features are using 4bits_bin, and the feature bundles will be created accordingly (line 371). 

If one feature only contains 0/1 values, but some other features have 255 bins, then max_num_bin_ will be set to 255, and dword_features_ will be set to 4. In this case, some features use 4bits_bin and some features use the 8-bit DenseBin. When creating the feature bundles, we have to check if each feature is using a 4bits_bin or not, as you can see around line 406. 

I actually found another small bug during testing this, which is fixed in `15c97b4c`.